For #2988. Added configuration for "Remember choice" checkbox in Permissions dialog

dector opened the pull request:
Please read #2988 and make diagonal-check solution in this PR. Is this a way-to-go solution?
Or should we go with breaking change instead (I suppose that it'll be nice to have ability to configure default behavior)?

### Pull Request checklist
- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

dector commented:
Ooops. Ment to make is draft PR. :unamused:

Amejia481 commented on components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsDialogFragment.kt:
```diff
@@ -178,6 +185,7 @@ internal class SitePermissionsDialogFragment : AppCompatDialogFragment() {
                 putInt(KEY_TITLE_ICON, titleIcon)
                 putBoolean(KEY_SHOULD_INCLUDE_CHECKBOX, shouldIncludeCheckBox)
+                putBoolean(KEY_SHOULD_SELECT_REMEMBER_CHOICE, shouldSelectRememberChoice)
                 putBoolean(KEY_USER_CHECK_BOX, shouldIncludeCheckBox)
```
We have to update the`KEY_USER_CHECK_BOX (userSelectionCheckBox)` with the value of `shouldSelectRememberChoice` because this is the property that stores the user decision. Right now it's using `shouldIncludeCheckBox` value if the user hits the positive button his/her decision will be store even thought the checkbox is not unchecked.

```kotlin
        positiveButton.setOnClickListener {
            feature?.onPositiveButtonPress(sessionId, userSelectionCheckBox)
            dismiss()
      }
```

Amejia481 replied:
`shouldIncludeCheckBox`  only gets updated when the user hit checkbox.
  
```kotlin
  containerView.findViewById<CheckBox>(R.id.do_not_ask_again).apply {
            visibility = VISIBLE
            isChecked = checked
            setOnCheckedChangeListener { _, isChecked ->
                userSelectionCheckBox = isChecked
            }
        }
```

Amejia481 replied:
We don't need anymore `KEY_USER_CHECK_BOX`

Amejia481 commented on components/feature/sitepermissions/src/test/java/mozilla/components/feature/sitepermissions/SitePermissionsDialogFragmentTest.kt:
```diff
@@ -53,6 +53,30 @@ class SitePermissionsDialogFragmentTest {
+
+        val checkBox = dialog.findViewById<CheckBox>(R.id.do_not_ask_again)
+
+        assertTrue(checkBox.isVisible())
```
I think we also have to check that `userSelectionCheckBox` gets initialize with the right value.

dector replied:
Good point.

dector commented on components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsDialogFragment.kt:
```diff
@@ -167,7 +173,8 @@ internal class SitePermissionsDialogFragment : AppCompatDialogFragment() {
             shouldIncludeCheckBox: Boolean,
-            isNotificationRequest: Boolean = false
+            isNotificationRequest: Boolean = false,
+            shouldSelectRememberChoice: Boolean = true
```
I agree.

Sorry, I'm not aware with internal engineering processes. But I suppose that the combination of two facts:
  - this component is in under-production (1.0) state;
  - you are ready to take responsibility for this change :smile:;
is enough in this case. Will fix it.

dector commented:
@Amejia481 done. Changes:
  - Rebased with `master`.
  - Added few more tests.
  - Documented public API.
  - Made `dialogConfig` nullable and `shouldPreselectDoNotAskAgain=false` by default.
  - Logically grouped arguments in the top of `SitePermissionsDialogFragment`.
  - Initializing `userSelectionCheckBox` in lateinit-way (as it's not argument but state).

dector commented:
> When we use 1) it ends up creating additional objects for just storing a primitive boolean. I think it doesn't worth the cost.
> Maybe this talk from I/O maybe interesting [Kotlin Under the Hood: Understand the Internals](https://www.youtube.com/watch?v=Ta5wBJsC39s) :)

Yep, I'm aware of hidden cost for delegation. Used it is here for few reasons:
  - `lateinit` isn't available for primitives.
  - Defining this variable with default value is more error-prone because it can be used without proper initialization (breaking fail-fast principle). [A]
  - [improvements in latest ART runtime](https://www.youtube.com/watch?v=Zc4JP8kNGmQ) soon be quite widespread. (Chet and Romain are really awesome speakers!).

However in the end of the day I think that I'm agree with your concern/compromise here. Base components should be as fast/efficient as possible and all ambiguity described in [A] should be heavily covered with tests.

Thanks a lot for review! :) :balloon:

Amejia481 commented:
You are more than welcome :) , keep up with the good job!
